Sources/Workspace/LocalContainerProvider.swift

```diff
@@ -0,0 +1,150 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import PackageModel
+import PackageGraph
+import TSCBasic
+import TSCUtility
+import SourceControl
+
+/// PackageContainerProvider implementation used by Workspace to do a dependency pre-calculation using the cached
+/// dependency information (Workspace.DependencyManifests) to check if dependency resolution is required before
+/// performing a full resolution.
+struct LocalContainerProvider: PackageContainerProvider {
+    /// The package graph inputs.
+    let root: PackageGraphRoot
+
+    /// The managed manifests to make available to the resolver.
+    let dependencyManifests: Workspace.DependencyManifests
+
+    /// The SwiftPM config.
+    let config: SwiftPMConfig
+
+    /// The tools version currently in use.
+    let currentToolsVersion: ToolsVersion
+
+    /// The diagnostics engine to report errors.
+    let diagnostics: DiagnosticsEngine
+
+    init(
+        root: PackageGraphRoot,
+        dependencyManifests: Workspace.DependencyManifests,
+        config: SwiftPMConfig,
+        currentToolsVersion: ToolsVersion = ToolsVersion.currentToolsVersion,
+        diagnostics: DiagnosticsEngine
+    ) {
+        self.root = root
+        self.dependencyManifests = dependencyManifests
+        self.config = config
+        self.currentToolsVersion = currentToolsVersion
+        self.diagnostics = diagnostics
+    }
+
+    func getContainer(
+        for identifier: PackageReference,
+        skipUpdate: Bool,
+        completion: @escaping (Result<PackageContainer, AnyError>) -> Void
+    ) {
+        // Start by searching manifests from the Workspace's resolved dependencies.
+        if let manifest = dependencyManifests.dependencies.first(where: { $1.packageRef == identifier }) {
+            let container = LocalPackageContainer(
+                manifest: manifest.manifest,
+                dependency: manifest.dependency,
+                config: config,
+                currentToolsVersion: currentToolsVersion
+            )
+
+            return completion(.success(container))
+        }
+
+        // Continue searching from the Workspace's root manifests.
+        if let index = dependencyManifests.root.packageRefs.firstIndex(of: identifier) {
+            let container = LocalPackageContainer(
+                manifest: dependencyManifests.root.manifests[index],
+                dependency: nil,
+                config: config,
+                currentToolsVersion: currentToolsVersion
+            )
+
+            return completion(.success(container))
+        }
+
+        // As we don't have anything else locally, error out.
+        completion(.init(Diagnostics.fatalError))
```

Comment:
does this ever happen?

Reply:
I think I remember this happens when you add a new dependency as it doesn't exist locally.

Reply:
ah.